### ISSUE TITLE
False Positive - ``assignment-from-none`` raised when function not necessarily returns `None`

### ISSUE THREAD

**Issue by orSolocate** [PyCQA/pylint#7853]
### Bug description

`assignment-from-none` is emitted where `func`  does not necessarily returns `None`

consider this `example.py` file to reproduce the error:

```
def get_func(param):

    if param is None:
        def func():
            return None
    else:
        def func():
            return param

    return func


def process_val(param):

    func = get_func(param)
    val = func()

    # Do stuff with 'val'; *None* is an useful case.
    return val
```

### Command used

```shell
pylint example.py
```

### Pylint output

```shell
`:16:4: E1128: Assigning result of a function call, where the function returns None (assignment-from-none)`
```

### Expected behavior

No Error

### Pylint version

```shell
pylint 2.16.0-dev
astroid 2.12.13
Python 3.8.10
```

### OS / Environment

Windows 10

**Comment by DanielNoord** [comment]
This is way too intricate for a static code analyser to ever understand... I don't think we'll ever be able to implement this and would vote for closing the issue.

**Comment by Pierre-Sassoulas** [comment]
I agree that we'll have a hard time handling this correctly. But the message level should be changed to a warning imo. Nothing is breaking if you assign from None. And astroid should raise an inference error if it can't infer correctly the return value.

**Comment by orSolocate** [comment]
I could try and debug it and will share what astroid says when encounters this pattern.
I agree with the message type change, also maybe the message itself could be modified a bit, maybe to something like:
`Assigning result of a function call, where the function **might** return None`

**Comment by nickdrozd** [comment]
The function that is returned from `get_func` is not a function that sometimes returns `None`; rather, sometimes it is a function that always returns `None` and sometimes it isn't. So it isn't clear to me that this is a false positive.

Actually, I would say there is a false *negative* at play here, because a minor variation of this code does not raise a warning:

```python
    if param is None:
        def func():
            return param
    else:
        def func():
            return None
```

I suppose we can all agree that there is something wrong with this check, but not exactly what!

For reference, here is the code that raises the warning:

```python
        return_nodes = list(
            function_node.nodes_of_class(nodes.Return, skip_klass=nodes.FunctionDef)
        )
        if not return_nodes:
            self.add_message("assignment-from-no-return", node=node)
        else:
            for ret_node in return_nodes:
                if not (
                    isinstance(ret_node.value, nodes.Const)
                    and ret_node.value.value is None
                    or ret_node.value is None
                ):
                    break
            else:
                self.add_message("assignment-from-none", node=node)
```

Look at that, some nice [mixed boolean operators](https://github.com/PyCQA/pylint/issues/2339) :laughing:

**Comment by Pierre-Sassoulas** [comment]
A general principle is that we're favoring false negatives over false positives as much as reasonably possible. False positive are a lot more troublesome than false negative for us as maintainers as well as for users. 

But on top of that, in this case I would be very wary of raising the message when the function is possibly not ``None`` some of the time. It means we're in a pattern used very often of returning an ``Optional[x]`` only with a convoluted approach using multiple function instead of multiple branches. The user probably knows what they're doing (they want the value when it's not None and are probably checking for noneness at some point). We have to be absolutely sure it's None for all possible branches or the check does not make sense. 

We're already checking for all branches of the code but we also need to check **all possible definition of the function** imo.

**Comment by Pierre-Sassoulas** [comment]
> The user would be declaring that they are intentionally defining a function that always explicitly returns None.

It's not an issue to declare a function that always return ``None``, ``print`` does that for example. The issue is doing ``x = print("Hello world")`` because then ``x`` will always be ``None`` so the assignment is useless and that's a code smell. ``assignment-from-none`` detects this code smell.

**Comment by orSolocate** [comment]
I investigated the issue, The problem is with the `safe_infer` function in the `utils.py` module. It infers the first `FunctionDef` and gets its `pytype`, then compared all the other `FunctionDefs` to this pytype. The `node.infer(context=context)` finds both function definitions in our case, but the `_get_python_type_of_node` returns the same value of `function.builtins` both for `param` and `None` returns. This explains @nickdrozd  reversed case, the returned node that is checked is the first one, which is `return param`, because it is not returning `None`, the message is not raised..

I modified the code and replaced the `_get_python_type_of_node` method with `infer_call_result` logic. It seems something like this:

for the first inferred FunctionDef:
```
        if isinstance(value, nodes.FunctionDef):
            try:
                x = value.infer_call_result(context=context)
                for type in x:
                    if type is astroid.Uninferable:
                        return None
                    inferred_types.add(type.pytype())
            except astroid.InferenceError:
                return None  # There is some kind of ambiguity
        else:
            inferred_types.add(_get_python_type_of_node(value))
```

for the second inferred FunctionDef:
```
            if isinstance(inferred, nodes.FunctionDef):
                inferred_type = inferred.infer_call_result(context=context)
                try:
                    for inf_val_type in inferred_type:
                        if inf_val_type not in inferred_types:
                            return None
                except astroid.InferenceError:
                    return None
            else:
                inferred_type = _get_python_type_of_node(inferred)
                if inferred_type not in inferred_types:
                    return None  # If there is ambiguity on the inferred node.
```

This code still has regression failure with 28 tests failing, not so bad and maybe could be fixed with improvements to my implementation.

is this really of great interest for us to fix or is this a safe fix?
should I make a MR for this?

**Comment by Pierre-Sassoulas** [comment]
Could you open a merge request so we can see the result of the test and the result on primer please @orSolocate ?

**Comment by DanielNoord** [comment]
Sorry for chiming in so late, why can't we fix `_get_python_type_of_node`? That function seems to be returning the same value for two different "types".

**Comment by orSolocate** [comment]
@DanielNoord  thanks for digging into this mate! 
the `safe_infer` used in the `assignment-from-none` checker is actually for the function node. so it is okay that the `_get_python_type_of_node` returns the same type for both function declarations in the example, a type which is a function.

The problem with the current code is that the `safe_infer` only returns the first "infer " of the function, while in our checker we want to check that they return `None` for all possible "infers".

Let's continue the discussion in the PR #7886.

**Comment by orSolocate** [comment]
Edit:
by "infers" I meant it returns one of these guys:
```
'\ndef func():\n    return None'

'\ndef func():\n    return param'

```

**Comment by jacobtylerwalls** [comment]
The assignment-from-none check needs to consider all possible inferred functions, not merely the first, so it should use `infer_all()`, not `safe_infer()`.
